packages/marshaller/src/ddl2/ddl.ts

```diff
@@ -30,24 +33,52 @@ class DDLDatasourceAdapter {
         this._dsDedup = {};
     }
 
-    set(ds: DDL2.DatasourceType) {
-        this._dsDedup[ds.id] = ds;
+    id(dsT: DatasourceType): string {
+        if (dsT instanceof WU) {
+            return `${dsT.url()}/${dsT.wuid()}`;
+        } else if (dsT instanceof RoxieService) {
+            return `${dsT.url()}/${dsT.querySet()}/${dsT.queryID()}`;
+        }
+        return dsT.id();
     }
 
-    get(ds: DatasourceType): DDL2.DatasourceType {
-        const dsID = ds.id();
-        let retVal: DDL2.DatasourceType = this._dsDedup[dsID];
-        if (!retVal) {
-            retVal = ds.toDDL();
-            this._dsDedup[dsID] = retVal;
+    append(pDS: DSPicker | DatasourceRefType) {
+        const dsT: DatasourceRefType = pDS instanceof DSPicker ? pDS.datasource() : pDS;
+        const dsTID = dsT.id();
+        const ds: DatasourceType = dsT instanceof RoxieResult ? dsT.service() : dsT instanceof WUResult ? dsT.wu() : dsT;
+        const dsID = this.id(ds);
+        if (!this._dsDedup[dsID]) {
+            this._dsDedup[dsID] = {
+                ds,
+                result: {}
+            };
+        } else {
+            //  Common up WU and RoxieService datasources!
+            if (dsT instanceof RoxieResult) {
+                dsT.service(this._dsDedup[dsID].ds as RoxieService);
+            } else if (dsT instanceof WUResult) {
+                dsT.wu(this._dsDedup[dsID].ds as WU);
+            }
+        }
+        if (!this._dsDedup[dsID].result[dsTID]) {
+            this._dsDedup[dsID].result[dsTID] = dsT;
         }
-        return retVal;
     }
 
     getAll(): DDL2.DatasourceType[] {
         const retVal: DDL2.DatasourceType[] = [];
         for (const key in this._dsDedup) {
-            retVal.push(this._dsDedup[key]);
+            const ddl = this._dsDedup[key].ds.toDDL();
+            if (ddl.type === "wuresult" || ddl.type === "roxie")
```

Comment:
This `if` statement should use curly brackets.

Reply:
Yup...